In [10]:
import pandas as pd
import datetime
import requests

Orion at NYU said this:

What you can do for now is use these links (current month daily, previous month daily, last three months weekly, previous three months weekly):

* [November](https://api.jaildatainitiative.org/dashboards/get_pops?jails=&races=&ages=&genders=&date=November)
* [October](https://api.jaildatainitiative.org/dashboards/get_pops?jails=&races=&ages=&genders=&date=October)
* [September - November](https://api.jaildatainitiative.org/dashboards/get_pops?jails=&races=&ages=&genders=&date=September-November)
* [June-August](https://api.jaildatainitiative.org/dashboards/get_pops?jails=&races=&ages=&genders=&date=June-August)

## Get New Data

In [3]:
def pull_jdi_api(url):
    response = requests.get(url)
    df = pd.DataFrame(response.json()['data'])
    df.Date = pd.to_datetime(df.Date)
    return df

In [5]:
url = "https://api.jaildatainitiative.org/dashboards/get_pops?jails=&races=&ages=&genders=&date=November"
response = requests.get(url)
nov = pd.DataFrame(response.json()['data'])
nov.head()

Date Race  Age Gender             JailId  \
0  2021-11-01T00:00:00.000Z  All  All    All    FL-Hillsborough   
1  2021-11-01T00:00:00.000Z  All  All    All          TX-Collin   
2  2021-11-01T00:00:00.000Z  All  All    All         MN-Stearns   
3  2021-11-01T00:00:00.000Z  All  All    All          IN-Marion   
4  2021-11-01T00:00:00.000Z  All  All    All  VA-Virginia_Beach   

   Population_Interpolated  Admissions_Interpolated  Releases_Interpolated  \
0                   2945.0                     48.0                    0.0   
1                   1148.0                     40.0                   36.0   
2                    153.0                      8.0                   18.0   
3                   2037.0                     64.0                  102.0   
4                   1190.0                     13.0                   17.0   

   ScrapeMissing  
0              0  
1              0  
2              0  
3              0  
4              0

In [16]:
pull_time = datetime.datetime.now().strftime("%m-%d-%Y-%-H%M%P")
filename = "October_Data_pulled_{}.csv".format(pull_time)
print("Saving to: ", filename)
nov.to_csv(filename,index=False)

In [6]:
nov.ScrapeMissing.value_counts(normalize=True)

0    0.959542
1    0.040458
Name: ScrapeMissing, dtype: float64

In [8]:
oct_url = "https://api.jaildatainitiative.org/dashboards/get_pops?jails=&races=&ages=&genders=&date=October"
october = pull_jdi_api(oct_url)
october.head(2)

Date Race  Age Gender           JailId  \
0 2021-10-01 00:00:00+00:00  All  All    All  FL-Hillsborough   
1 2021-10-01 00:00:00+00:00  All  All    All        TX-Collin   

   Population_Interpolated  Admissions_Interpolated  Releases_Interpolated  \
0              2867.000000                 90.00000               0.000000   
1              1113.978261                 10.73913              11.282609   

   ScrapeMissing  
0              0  
1              1

In [15]:
pull_time = datetime.datetime.now().strftime("%m-%d-%Y-%-H%M%P")
filename = "October_Data_pulled_{}.csv".format(pull_time)
print("Saving to: ", filename)
october.to_csv(filename,index=False)

In [17]:
october.ScrapeMissing.value_counts(normalize=True)

0    0.927201
1    0.072799
Name: ScrapeMissing, dtype: float64

In [19]:
sept_nov_wkly_url = "https://api.jaildatainitiative.org/dashboards/get_pops?jails=&races=&ages=&genders=&date=September-November"
sept_nov_wkly = pull_jdi_api(sept_nov_wkly_url)
sept_nov_wkly.head(2)

Date Race  Age Gender           JailId  \
0 2021-08-29 00:00:00+00:00  All  All    All  FL-Hillsborough   
1 2021-08-29 00:00:00+00:00  All  All    All        TX-Collin   

   Population_Interpolated  Admissions_Interpolated  Releases_Interpolated  \
0              2821.250000                 67.75000              56.500000   
1              1129.467391                 10.73913              11.282609   

   ScrapeMissing  
0              0  
1              1

In [20]:
pull_time = datetime.datetime.now().strftime("%m-%d-%Y-%-H%M%P")
filename = "Sept-Nov_Weekly_Data_pulled_{}.csv".format(pull_time)
print("Saving to: ", filename)
sept_nov_wkly.to_csv(filename,index=False)

Saving to:  Sept-Nov_Weekly_Data_pulled_01-09-2022-1033am.csv


In [24]:
june_aug_wkly_url = "https://api.jaildatainitiative.org/dashboards/get_pops?jails=&races=&ages=&genders=&date=June-August"
june_aug_wkly = pull_jdi_api(june_aug_wkly_url)

pull_time = datetime.datetime.now().strftime("%m-%d-%Y-%-H%M%P")
filename = "June-Aug_Weekly_Data_pulled_{}.csv".format(pull_time)
print("Saving to: ", filename)
june_aug_wkly.to_csv(filename,index=False)

june_aug_wkly.head(2)

Saving to:  June-Aug_Weekly_Data_pulled_01-09-2022-1044am.csv


Date Race  Age Gender     JailId  \
0 2021-05-30 00:00:00+00:00  All  All    All  TX-Collin   
1 2021-05-30 00:00:00+00:00  All  All    All  IN-Marion   

   Population_Interpolated  Admissions_Interpolated  Releases_Interpolated  \
0                    980.4                     43.4                   37.6   
1                   1906.2                     79.4                   86.6   

   ScrapeMissing  
0              0  
1              0

## Load Previous Output

In [29]:
prior = pd.read_excel("Jail_Summaries_as_of_10-08-2021.xlsx", engine="openpyxl")
print(prior.shape)
prior.head(3)

(415, 79)


STATE-COUNTY  2020-03-10 00:00:00  2020-03-16 00:00:00  2020-04-06 00:00:00  \
0   AL-Autauga                  172                178.0                139.0   
1   AL-Chilton                  212                207.0                152.0   
2      AL-Clay                   39                 42.0                 31.0   

   2020-05-04 00:00:00  2020-06-01 00:00:00  2020-07-06 00:00:00  \
0                156.0                174.0                170.0   
1                172.0                175.0                184.0   
2                 28.0                 31.0                 34.0   

   2020-08-03 00:00:00  2020-09-07 00:00:00  2020-10-05 00:00:00  ...  \
0                147.0                160.0                160.0  ...   
1                158.0                153.0                188.0  ...   
2                 30.0                 32.0                 36.0  ...   

   FirstMondayAvailable07-2020  FirstMondayAvailable07-2021  \
0                        170.0                        158.0   
1                        184.0                        198.0   
2                         34.0                         60.0   

   FirstMondayAvailable08-2020  FirstMondayAvailable08-2021  \
0                        147.0                        159.0   
1                        158.0                        216.0   
2                         30.0                         44.0   

   FirstMondayAvailable09-2020  FirstMondayAvailable09-2021  \
0                        160.0                        150.0   
1                        153.0                        194.0   
2                         32.0                         54.0   

   FirstMondayAvailable10-2020  FirstMondayAvailable10-2021  \
0                        160.0                        158.0   
1                        188.0                        218.0   
2                         36.0                         54.0   

   FirstMondayAvailable11-2020  FirstMondayAvailable12-2020  
0                        185.0                        184.0  
1                        184.0                        200.0  
2                         33.0                         30.0  

[3 rows x 79 columns]

In [30]:
prior_sheet2 = pd.read_excel("Jail_Summaries_as_of_10-08-2021.xlsx", sheet_name=1, engine="openpyxl")
print(prior_sheet2.shape)
prior_sheet2.head(3)

(577, 4)


Scrape_Date  Count_Of_Jails  Total_Jail_Population  \
0  2020-03-10             415                 116372   
1  2020-03-11             399                 112775   
2  2020-03-12             404                 114083   

   Seven_Day_Rolling_Average  
0                        NaN  
1                        NaN  
2                        NaN

In [31]:
prior_sheet2.tail(3)

Scrape_Date  Count_Of_Jails  Total_Jail_Population  \
574  2021-10-05             397                  99014   
575  2021-10-06             396                  98615   
576  2021-10-07             394                  97997   

     Seven_Day_Rolling_Average  
574               98766.857143  
575               98704.571429  
576               98643.571429

In [32]:
prior.columns

Index([               'STATE-COUNTY',           2020-03-10 00:00:00,
                 2020-03-16 00:00:00,           2020-04-06 00:00:00,
                 2020-05-04 00:00:00,           2020-06-01 00:00:00,
                 2020-07-06 00:00:00,           2020-08-03 00:00:00,
                 2020-09-07 00:00:00,           2020-10-05 00:00:00,
                 2020-11-02 00:00:00,           2020-12-07 00:00:00,
                 2021-01-04 00:00:00,           2021-02-01 00:00:00,
                 2021-03-01 00:00:00,           2021-04-05 00:00:00,
                 2021-05-03 00:00:00,           2021-06-07 00:00:00,
                 2021-07-05 00:00:00,           2021-07-16 00:00:00,
                 2021-08-01 00:00:00,           2021-08-02 00:00:00,
                 2021-08-16 00:00:00,           2021-08-19 00:00:00,
                 2021-08-23 00:00:00,           2021-08-26 00:00:00,
                 2021-08-27 00:00:00,           2021-08-28 00:00:00,
                 2021-08-29 00:00:

## Questions 

* Is Weekly an Average of Daily?
* Do JailIds align with previous output file?

## Do JailIds Match to Prior File STATE-COUNTY?

In [36]:
nov.JailId.isin(prior['STATE-COUNTY']).value_counts(normalize=True)

False    0.594347
True     0.405653
Name: JailId, dtype: float64

### 96% of entries found for November.

In [37]:
prior['STATE-COUNTY'].isin(nov.JailId).value_counts(normalize=True)

True     0.975904
False    0.024096
Name: STATE-COUNTY, dtype: float64

### 97.5% Found for October

In [38]:
prior['STATE-COUNTY'].isin(october.JailId).value_counts(normalize=True)

True     0.985542
False    0.014458
Name: STATE-COUNTY, dtype: float64

## Jails in Nov VS Oct Daily

In [39]:
print("Percentage of Nov Jails in Oct File:")
print(nov.JailId.isin(october.JailId).value_counts(normalize=True))
print("Nov Ids not in Oct:")
nov[~nov.JailId.isin(october.JailId)]['JailId'].unique()

Percentage of Nov Jails in Oct File:
True     0.981959
False    0.018041
Name: JailId, dtype: float64
Nov Ids not in Oct:


array(['TX-Kerr', 'AR-Randolph', 'FL-Gadsden', 'LA-Plaquemines',
       'TX-Comal', 'TX-Grayson', 'FL-Columbia', 'TX-Gregg', 'TX-Parker',
       'FL-Hamilton', 'TX-Henderson', 'TX-Hale', 'TX-Howard', 'TX-Lamar',
       'TX-Wood', 'TX-Andrews', 'MN-Benton', 'AR-Izard'], dtype=object)

In [40]:
print("Percentage of Oct Jails in Nov File:")
print(october.JailId.isin(nov.JailId).value_counts(normalize=True))

print("Oct Ids not in Nov:")
october[~october.JailId.isin(nov.JailId)]['JailId'].unique()

Percentage of Oct Jails in Nov File:
True     0.988893
False    0.011107
Name: JailId, dtype: float64
Oct Ids not in Nov:


array(['AZ-Pinal', 'CA-Shasta', 'WI-Milwaukee', 'WI-Ozaukee',
       'TX-Wichita', 'KY-Greenup', 'MA-Worcester', 'MT-Yellowstone',
       'NM-Hobbs', 'MN-Cook', 'ID-Canyon'], dtype=object)

## Full Summary

In [41]:
jails = set()
for df in [nov, october, sept_nov_wkly, june_aug_wkly]:
    jails = jails.union(set(df.JailId.unique()))
jail_summaries = pd.DataFrame(sorted(list(jails)))
jail_summaries.columns = ['JailId']
# jail_summaries.set_index('JailId')
jail_summaries['Nov'] = jail_summaries['JailId'].isin(nov.JailId.unique())
jail_summaries['Oct'] = jail_summaries['JailId'].isin(october.JailId.unique())
jail_summaries['Sept-Nov'] = jail_summaries['JailId'].isin(sept_nov_wkly.JailId.unique())
jail_summaries['June-Aug'] = jail_summaries['JailId'].isin(june_aug_wkly.JailId.unique())
jail_summaries.to_csv("JDI_Coverage_By_JailID_and_ScrapeFile.csv", index=False)
jail_summaries.head()

JailId   Nov   Oct  Sept-Nov  June-Aug
0   AL-Autauga  True  True      True      True
1   AL-Baldwin  True  True      True      True
2    AL-Blount  True  True      True      True
3   AL-Calhoun  True  True      True      True
4  AL-Cherokee  True  True      True      True

## Nov VS Prior File

In [42]:
missing = prior[~prior['STATE-COUNTY'].isin(nov.JailId)]['STATE-COUNTY'].unique()
print(len(missing))
missing

10


array(['CA-Shasta', 'ID-Canyon', 'KY-Allen', 'MA-Worcester', 'MN-Brown',
       'NE-Lancaster', 'NM-Hobbs', 'OR-Marion_Work_Center', 'TX-Brown',
       'TX-Lubbock'], dtype=object)

In [43]:
from fuzzywuzzy import fuzz, process

/home/matt/.local/lib/python3.6/site-packages/fuzzywuzzy/fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')


In [44]:
process.extractBests('MA-Worcester', nov[nov.JailId.str.startswith("MA-")].JailId.unique())

[]

### November Does Not Contain MA Jails.

In [45]:
nov[nov.JailId.str.startswith("MA-")]

Empty DataFrame
Columns: [Date, Race, Age, Gender, JailId, Population_Interpolated, Admissions_Interpolated, Releases_Interpolated, ScrapeMissing]
Index: []

#### Does Exist in Oct File

In [46]:
october[october.JailId.str.startswith("MA-")].head()

Date Race  Age Gender        JailId  \
746  2021-10-01 00:00:00+00:00  All  All    All  MA-Worcester   
1732 2021-10-02 00:00:00+00:00  All  All    All  MA-Worcester   
2722 2021-10-03 00:00:00+00:00  All  All    All  MA-Worcester   
3710 2021-10-04 00:00:00+00:00  All  All    All  MA-Worcester   
4699 2021-10-05 00:00:00+00:00  All  All    All  MA-Worcester   

      Population_Interpolated  Admissions_Interpolated  Releases_Interpolated  \
746                     641.0                     19.0                   15.0   
1732                    633.0                      7.0                    2.0   
2722                    634.0                      3.0                    1.0   
3710                    633.0                      0.0                    2.0   
4699                    641.0                     10.0                   15.0   

      ScrapeMissing  
746               0  
1732              0  
2722              0  
3710              0  
4699              0

## Prior Missing From October

In [47]:
missing = prior[~prior['STATE-COUNTY'].isin(october.JailId)]['STATE-COUNTY'].unique()
print(len(missing))
missing

6


array(['KY-Allen', 'MN-Brown', 'NE-Lancaster', 'OR-Marion_Work_Center',
       'TX-Brown', 'TX-Lubbock'], dtype=object)

## Prior Missing From October and Nov Daily

In [48]:
missing = prior[(~prior['STATE-COUNTY'].isin(nov.JailId))
               & (~prior['STATE-COUNTY'].isin(october.JailId))]['STATE-COUNTY'].unique()
print(len(missing))
missing

6


array(['KY-Allen', 'MN-Brown', 'NE-Lancaster', 'OR-Marion_Work_Center',
       'TX-Brown', 'TX-Lubbock'], dtype=object)

## Prior Missing From All

In [49]:
missing = prior[(~prior['STATE-COUNTY'].isin(nov.JailId))
               & (~prior['STATE-COUNTY'].isin(october.JailId))
               & (~prior['STATE-COUNTY'].isin(sept_nov_wkly.JailId))
               & (~prior['STATE-COUNTY'].isin(june_aug_wkly.JailId))]['STATE-COUNTY'].unique()
print(len(missing))
missing

3


array(['NE-Lancaster', 'OR-Marion_Work_Center', 'TX-Lubbock'],
      dtype=object)

## Daily vs Weekly Comparisons

In [50]:
ex = "AL-Autauga"
nov[nov.JailId == "AL-Autauga"].head(20)

Date Race  Age Gender      JailId  \
114    2021-11-01T00:00:00.000Z  All  All    All  AL-Autauga   
1113   2021-11-02T00:00:00.000Z  All  All    All  AL-Autauga   
2110   2021-11-03T00:00:00.000Z  All  All    All  AL-Autauga   
3110   2021-11-04T00:00:00.000Z  All  All    All  AL-Autauga   
4108   2021-11-05T00:00:00.000Z  All  All    All  AL-Autauga   
5103   2021-11-06T00:00:00.000Z  All  All    All  AL-Autauga   
6102   2021-11-07T00:00:00.000Z  All  All    All  AL-Autauga   
7097   2021-11-08T00:00:00.000Z  All  All    All  AL-Autauga   
8098   2021-11-09T00:00:00.000Z  All  All    All  AL-Autauga   
9095   2021-11-10T00:00:00.000Z  All  All    All  AL-Autauga   
10091  2021-11-11T00:00:00.000Z  All  All    All  AL-Autauga   
11088  2021-11-12T00:00:00.000Z  All  All    All  AL-Autauga   
12085  2021-11-13T00:00:00.000Z  All  All    All  AL-Autauga   
13084  2021-11-14T00:00:00.000Z  All  All    All  AL-Autauga   
14085  2021-11-15T00:00:00.000Z  All  All    All  AL-Autauga   
15085  2021-11-16T00:00:00.000Z  All  All    All  AL-Autauga   
16079  2021-11-17T00:00:00.000Z  All  All    All  AL-Autauga   
17073  2021-11-18T00:00:00.000Z  All  All    All  AL-Autauga   
18071  2021-11-19T00:00:00.000Z  All  All    All  AL-Autauga   
19067  2021-11-20T00:00:00.000Z  All  All    All  AL-Autauga   

       Population_Interpolated  Admissions_Interpolated  \
114                      162.0                      2.0   
1113                     165.0                      4.0   
2110                     165.0                      4.0   
3110                     162.0                      4.0   
4108                     160.0                      4.0   
5103                     159.0                      1.0   
6102                     162.0                      4.0   
7097                     161.0                      1.0   
8098                     161.0                      1.0   
9095                     165.0                      8.0   
10091                    176.0                     11.0   
11088                    177.0                      2.0   
12085                    173.0                      0.0   
13084                    177.0                      6.0   
14085                    176.0                      3.0   
15085                    171.0                      3.0   
16079                    181.0                     11.0   
17073                    178.0                      3.0   
18071                    168.0                      2.0   
19067                    166.0                      8.0   

       Releases_Interpolated  ScrapeMissing  
114                      0.0              0  
1113                     0.0              0  
2110                     0.0              0  
3110                     0.0              0  
4108                     0.0              0  
5103                     0.0              0  
6102                     0.0              0  
7097                     0.0              0  
8098                     0.0              0  
9095                     0.0              0  
10091                    0.0              0  
11088                    0.0              0  
12085                    0.0              0  
13084                    0.0              0  
14085                    0.0              0  
15085                    0.0              0  
16079                    0.0              0  
17073                    0.0              0  
18071                    0.0              0  
19067                    0.0              0

In [51]:
nov.Date = pd.to_datetime(nov.Date)

In [52]:
nov[nov.JailId == "AL-Autauga"].set_index('Date').rolling(7).mean()#.groupby(pd.Grouper(freq="W")).max()

Population_Interpolated  Admissions_Interpolated  \
Date                                                                          
2021-11-01 00:00:00+00:00                      NaN                      NaN   
2021-11-02 00:00:00+00:00                      NaN                      NaN   
2021-11-03 00:00:00+00:00                      NaN                      NaN   
2021-11-04 00:00:00+00:00                      NaN                      NaN   
2021-11-05 00:00:00+00:00                      NaN                      NaN   
2021-11-06 00:00:00+00:00                      NaN                      NaN   
2021-11-07 00:00:00+00:00               162.142857                 3.285714   
2021-11-08 00:00:00+00:00               162.000000                 3.142857   
2021-11-09 00:00:00+00:00               161.428571                 2.714286   
2021-11-10 00:00:00+00:00               161.428571                 3.285714   
2021-11-11 00:00:00+00:00               163.428571                 4.285714   
2021-11-12 00:00:00+00:00               165.857143                 4.000000   
2021-11-13 00:00:00+00:00               167.857143                 3.857143   
2021-11-14 00:00:00+00:00               170.000000                 4.142857   
2021-11-15 00:00:00+00:00               172.142857                 4.428571   
2021-11-16 00:00:00+00:00               173.571429                 4.714286   
2021-11-17 00:00:00+00:00               175.857143                 5.142857   
2021-11-18 00:00:00+00:00               176.142857                 4.000000   
2021-11-19 00:00:00+00:00               174.857143                 4.000000   
2021-11-20 00:00:00+00:00               173.857143                 5.142857   
2021-11-21 00:00:00+00:00               172.428571                 4.857143   
2021-11-22 00:00:00+00:00               170.714286                 4.714286   
2021-11-23 00:00:00+00:00               169.714286                 4.714286   
2021-11-24 00:00:00+00:00               167.142857                 3.857143   
2021-11-25 00:00:00+00:00               164.857143                 3.714286   
2021-11-26 00:00:00+00:00               164.285714                 3.571429   
2021-11-27 00:00:00+00:00               164.000000                 2.714286   
2021-11-28 00:00:00+00:00               164.000000                 2.571429   
2021-11-29 00:00:00+00:00               164.142857                 2.428571   
2021-11-30 00:00:00+00:00               164.285714                 2.428571   

                           Releases_Interpolated  ScrapeMissing  
Date                                                             
2021-11-01 00:00:00+00:00                    NaN            NaN  
2021-11-02 00:00:00+00:00                    NaN            NaN  
2021-11-03 00:00:00+00:00                    NaN            NaN  
2021-11-04 00:00:00+00:00                    NaN            NaN  
2021-11-05 00:00:00+00:00                    NaN            NaN  
2021-11-06 00:00:00+00:00                    NaN            NaN  
2021-11-07 00:00:00+00:00                    0.0            0.0  
2021-11-08 00:00:00+00:00                    0.0            0.0  
2021-11-09 00:00:00+00:00                    0.0            0.0  
2021-11-10 00:00:00+00:00                    0.0            0.0  
2021-11-11 00:00:00+00:00                    0.0            0.0  
2021-11-12 00:00:00+00:00                    0.0            0.0  
2021-11-13 00:00:00+00:00                    0.0            0.0  
2021-11-14 00:00:00+00:00                    0.0            0.0  
2021-11-15 00:00:00+00:00                    0.0            0.0  
2021-11-16 00:00:00+00:00                    0.0            0.0  
2021-11-17 00:00:00+00:00                    0.0            0.0  
2021-11-18 00:00:00+00:00                    0.0            0.0  
2021-11-19 00:00:00+00:00                    0.0            0.0  
2021-11-20 00:00:00+00:00                    0.0            0.0  
2021-11-21 00:00:00+00:00                    0

In [53]:
nov[nov.JailId == "AL-Autauga"].mean()

/home/matt/.local/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: DataFrame.mean and DataFrame.median with numeric_only=None will include datetime64 and datetime64tz columns in a future version.
  """Entry point for launching an IPython kernel.


Population_Interpolated    167.0
Admissions_Interpolated      3.6
Releases_Interpolated        0.0
ScrapeMissing                0.0
dtype: float64

In [54]:
temp = sept_nov_wkly[(sept_nov_wkly.JailId == "AL-Autauga")
             & (sept_nov_wkly.Date >= pd.to_datetime("2021-11-01", utc=True))]
temp.mean()

/home/matt/.local/lib/python3.6/site-packages/ipykernel_launcher.py:3: FutureWarning: DataFrame.mean and DataFrame.median with numeric_only=None will include datetime64 and datetime64tz columns in a future version.
  This is separate from the ipykernel package so we can avoid doing imports until


Population_Interpolated    167.845238
Admissions_Interpolated      3.511905
Releases_Interpolated        0.000000
ScrapeMissing                0.000000
dtype: float64

## Ability to Integrate with Prior File

#### Combine October and November File - check that total length of first two files matches that of joined file 

> **(used .shape which returns nrows, ncols)**

In [117]:
print(len(october)+len(nov))
nov_oct = pd.concat([nov, october])
print(nov_oct.shape)
nov_oct.head()

60633
(60633, 9)


Date Race  Age Gender             JailId  \
0 2021-11-01 00:00:00+00:00  All  All    All    FL-Hillsborough   
1 2021-11-01 00:00:00+00:00  All  All    All          TX-Collin   
2 2021-11-01 00:00:00+00:00  All  All    All         MN-Stearns   
3 2021-11-01 00:00:00+00:00  All  All    All          IN-Marion   
4 2021-11-01 00:00:00+00:00  All  All    All  VA-Virginia_Beach   

   Population_Interpolated  Admissions_Interpolated  Releases_Interpolated  \
0                   2945.0                     48.0                    0.0   
1                   1148.0                     40.0                   36.0   
2                    153.0                      8.0                   18.0   
3                   2037.0                     64.0                  102.0   
4                   1190.0                     13.0                   17.0   

   ScrapeMissing  
0              0  
1              0  
2              0  
3              0  
4              0

## Identifying Two MidPoints to Divide Time into Thirds

In [118]:
most_recent = pd.to_datetime("11-30-2021")
mar_10 = pd.to_datetime("03-10-2020")
day_length = (most_recent - mar_10).days
day_length

630

In [119]:
day_length / 3

210.0

In [120]:
first_tripoint = mar_10 + pd.Timedelta(days=day_length/3)
first_tripoint

Timestamp('2020-10-06 00:00:00')

In [121]:
second_tripoint = mar_10 + pd.Timedelta(days=(day_length/3*2))
second_tripoint

Timestamp('2021-05-04 00:00:00')

In [122]:
nov.head(2)

Date Race  Age Gender           JailId  \
0 2021-11-01 00:00:00+00:00  All  All    All  FL-Hillsborough   
1 2021-11-01 00:00:00+00:00  All  All    All        TX-Collin   

   Population_Interpolated  Admissions_Interpolated  Releases_Interpolated  \
0                   2945.0                     48.0                    0.0   
1                   1148.0                     40.0                   36.0   

   ScrapeMissing  
0              0  
1              0

In [123]:
nov.groupby(['JailId']).tail(1).Date.value_counts()

2021-11-30 00:00:00+00:00    998
Name: Date, dtype: int64

def parse_first_mondays(df):
    df['DayOfWeek'] = df.Scrape_Date.dt.strftime('%A')
    mondays = df[df.DayOfWeek=='Monday']
    first_mondays = mondays.groupby(['STATE-COUNTY', 'Month', 'Year']).head(1)
    return df

## 3 hr timecheck

## Format of DF to replicate

> See 
  line 115 of COVID_Prison_Transformations.py tagged at v1.0  
  or   
  output [13] from COVID_Prison_Transformations.ipynb tagged at v1.0


| County	| Population	| Scrape_Date |	State	| Facility_Identifier | 
|:---:|:---:|:---:|:---:|:---:|
|Autauga|	171|	2020-03-10|	AL|	AL-Autauga|
|Autauga|	172|	2020-03-11|	AL|	AL-Autauga|
|Autauga|	173|	2020-03-12|	AL|	AL-Autauga|
|Autauga|	179|	2020-03-13|	AL|	AL-Autauga|
|Autauga|	180|	2020-03-14|	AL|	AL-Autauga|

In [125]:
nov_oct.head()

Date Race  Age Gender             JailId  \
0 2021-11-01 00:00:00+00:00  All  All    All    FL-Hillsborough   
1 2021-11-01 00:00:00+00:00  All  All    All          TX-Collin   
2 2021-11-01 00:00:00+00:00  All  All    All         MN-Stearns   
3 2021-11-01 00:00:00+00:00  All  All    All          IN-Marion   
4 2021-11-01 00:00:00+00:00  All  All    All  VA-Virginia_Beach   

   Population_Interpolated  Admissions_Interpolated  Releases_Interpolated  \
0                   2945.0                     48.0                    0.0   
1                   1148.0                     40.0                   36.0   
2                    153.0                      8.0                   18.0   
3                   2037.0                     64.0                  102.0   
4                   1190.0                     13.0                   17.0   

   ScrapeMissing  
0              0  
1              0  
2              0  
3              0  
4              0

#### Data Quality Checks

In [126]:
nov_oct.Race.value_counts()

All    60633
Name: Race, dtype: int64

In [127]:
nov_oct.Age.value_counts()

All    60633
Name: Age, dtype: int64

In [128]:
nov_oct.Gender.value_counts()

All    60633
Name: Gender, dtype: int64

In [129]:
#Ensure unique Date / JailId Records
temp = nov_oct.groupby(['Date','JailId'])['Race'].count()
temp[temp!=1].head()

Series([], Name: Race, dtype: int64)

#### Rename Columns

In [130]:
rename_dict = {"Date":"Scrape_Date",
                        "JailId":"STATE-COUNTY",
                        "Population_Interpolated":"Population"}
nov_oct = nov_oct.rename(columns=rename_dict)
nov_oct = nov_oct[rename_dict.values()]

#### Create County and State Columns 

In [131]:
nov_oct['State'] = nov_oct["STATE-COUNTY"].apply(lambda x: x[:2])
nov_oct['County'] = nov_oct["STATE-COUNTY"].apply(lambda x: x[3:])

In [132]:
nov_oct

Scrape_Date       STATE-COUNTY  Population State  \
0     2021-11-01 00:00:00+00:00    FL-Hillsborough      2945.0    FL   
1     2021-11-01 00:00:00+00:00          TX-Collin      1148.0    TX   
2     2021-11-01 00:00:00+00:00         MN-Stearns       153.0    MN   
3     2021-11-01 00:00:00+00:00          IN-Marion      2037.0    IN   
4     2021-11-01 00:00:00+00:00  VA-Virginia_Beach      1190.0    VA   
...                         ...                ...         ...   ...   
30696 2021-10-31 00:00:00+00:00         AR-Madison         2.4    AR   
30697 2021-10-31 00:00:00+00:00             IA-Ida        16.0    IA   
30698 2021-10-31 00:00:00+00:00        OH-Fairborn         2.0    OH   
30699 2021-10-31 00:00:00+00:00         WY-Lincoln        22.0    WY   
30700 2021-10-31 00:00:00+00:00        WA-Columbia         8.0    WA   

               County  
0        Hillsborough  
1              Collin  
2             Stearns  
3              Marion  
4      Virginia_Beach  
...               ...  
30696         Madison  
30697             Ida  
30698        Fairborn  
30699         Lincoln  
30700        Columbia  

[60633 rows x 5 columns]

In [133]:
nov_oct["STATE-COUNTY"].nunique()

1009

## Join With Previous

#### Read Previous Raw Checkpoint Midscript

In [160]:
#See line 123 where exported from COVID_Prison_Transformations.py
df = pd.read_csv("RawData_03-10-21_to_10-07-21_Generated_01-09-2022.csv")
df.Scrape_Date = pd.to_datetime(df.Scrape_Date)
print(df.shape)
df.head()

(238731, 7)


Unnamed: 0  Population Scrape_Date STATE-COUNTY  Month  Year  DayOfWeek
0          31       172.0  2020-03-10   AL-Autauga      3  2020    Tuesday
1          32       173.0  2020-03-11   AL-Autauga      3  2020  Wednesday
2          33       174.0  2020-03-12   AL-Autauga      3  2020   Thursday
3          34       180.0  2020-03-13   AL-Autauga      3  2020     Friday
4          35       180.0  2020-03-14   AL-Autauga      3  2020   Saturday

In [161]:
prior.shape

(415, 79)

In [162]:
df['STATE-COUNTY'].nunique()

415

## Additional Checks on Previous Outputs

In [ ]:
# temp = pd.read_excel("Jail_Summaries_as_of_06-18-2021.xlsx", engine="openpyxl") #(565, 74)
# temp = pd.read_excel("Jail_Summaries_as_of_01-21-2021.xlsx", engine="openpyxl") #((429, 68)
# temp = pd.read_excel("Jail_Summaries_as_of_09-29-2021.xlsx", engine="openpyxl") #(590, 80)
# temp = pd.read_excel("Jail_Summaries_as_of_10-08-2021.xlsx", engine="openpyxl") #(415, 79)
# temp = pd.read_excel("Jail_Summaries_as_of_12-07-2020.xlsx", engine="openpyxl") #(598, 66)
# temp = pd.read_excel("Jail_Summaries_as_of_12-09-2020.xlsx", engine="openpyxl") #(598, 67)
# temp = pd.read_excel("Jail_Summaries_2020-03-10_to_2021-10-07_Generated_01-09-2022.xlsx", engine="openpyxl") #(598, 67)

# print(temp.shape)
# temp.head(2)

### Final QAs

> API Nov/Oct File Prepped is missing 6 Jails from the Previous Run on Oct 7th.

In [135]:
nov_oct[nov_oct['STATE-COUNTY'].isin(df['STATE-COUNTY'].unique())]['STATE-COUNTY'].nunique()

409

In [137]:
missing_from_api = df[~df['STATE-COUNTY'].isin(nov_oct['STATE-COUNTY'].unique())]['STATE-COUNTY'].unique()
print("{} Jails in Oct 7th File Missing from API.".format(len(missing_from_api)))
print(missing_from_api)

6 Jails in Oct 7th File Missing from API.
['KY-Allen' 'MN-Brown' 'NE-Lancaster' 'OR-Marion_Work_Center' 'TX-Brown'
 'TX-Lubbock']


#### Check Schemas

In [138]:
print(nov_oct.columns, df.columns)

Index(['Scrape_Date', 'STATE-COUNTY', 'Population', 'State', 'County'], dtype='object') Index(['Unnamed: 0', 'Population', 'Scrape_Date', 'STATE-COUNTY', 'Month',
       'Year', 'DayOfWeek'],
      dtype='object')


## Join Pieces

> Will Join Pieces First using `pd.concat` and then recalculate calculated fields:


```python
## Create Month / Day of Week / Year Columns 
df['Month'] = df.Scrape_Date.dt.strftime('%m')
df['Year'] = df.Scrape_Date.dt.strftime('%Y')
df['DayOfWeek'] = df.Scrape_Date.dt.strftime('%A')
#Convert Datetime to Strict Date
df['Scrape_Date'] = df['Scrape_Date'].dt.date
```

In [163]:
print(df.shape)
df.head(2)

(238731, 7)


Unnamed: 0  Population Scrape_Date STATE-COUNTY  Month  Year  DayOfWeek
0          31       172.0  2020-03-10   AL-Autauga      3  2020    Tuesday
1          32       173.0  2020-03-11   AL-Autauga      3  2020  Wednesday

In [142]:
[col for col in df.columns if col in nov_oct.columns]

['Population', 'Scrape_Date', 'STATE-COUNTY']

In [158]:
nov_oct.Scrape_Date = pd.to_datetime(nov_oct.Scrape_Date)
nov_oct['Scrape_Date'] = nov_oct['Scrape_Date'].dt.date

In [159]:
nov_oct.head(2)

Scrape_Date     STATE-COUNTY  Population State        County
0  2021-11-01  FL-Hillsborough      2945.0    FL  Hillsborough
1  2021-11-01        TX-Collin      1148.0    TX        Collin

In [153]:
print(len(df)+len(nov_oct))

299364


##### Check before overriding DF

In [154]:
temp = pd.concat([df, nov_oct])
print(len(temp))
temp.head(2)

299364


Unnamed: 0  Population          Scrape_Date STATE-COUNTY  Month    Year  \
0        31.0       172.0  2020-03-10 00:00:00   AL-Autauga    3.0  2020.0   
1        32.0       173.0  2020-03-11 00:00:00   AL-Autauga    3.0  2020.0   

   DayOfWeek State County  
0    Tuesday   NaN    NaN  
1  Wednesday   NaN    NaN

In [164]:
df = pd.concat([df, nov_oct])

df = df[['Population', 'Scrape_Date', 'STATE-COUNTY']] #Subset columns


print(df.shape)
df.head(3)

(299364, 3)


Population          Scrape_Date STATE-COUNTY
0       172.0  2020-03-10 00:00:00   AL-Autauga
1       173.0  2020-03-11 00:00:00   AL-Autauga
2       174.0  2020-03-12 00:00:00   AL-Autauga

In [165]:
## Create Month / Day of Week / Year Columns 
df.Scrape_Date = pd.to_datetime(df.Scrape_Date)

df['Month'] = df.Scrape_Date.dt.strftime('%m')
df['Year'] = df.Scrape_Date.dt.strftime('%Y')
df['DayOfWeek'] = df.Scrape_Date.dt.strftime('%A')
#Convert Datetime to Strict Date
df['Scrape_Date'] = df['Scrape_Date'].dt.date

## Save Checkpoint to Disk

In [166]:
df.head(2)

Population Scrape_Date STATE-COUNTY Month  Year  DayOfWeek
0       172.0  2020-03-10   AL-Autauga    03  2020    Tuesday
1       173.0  2020-03-11   AL-Autauga    03  2020  Wednesday

In [167]:
df.to_csv("Combined_Prior_to_Full_Script_Run_01-09-2022.csv", index=False)

## Reapply Rest of Script

## Appendix 5hr timecheck

## DevOps - environment recreation, if needed

In [70]:
import sys

sys.path

['/usr/lib/python36.zip',
 '/usr/lib/python3.6',
 '/usr/lib/python3.6/lib-dynload',
 '',
 '/home/matt/.local/lib/python3.6/site-packages',
 '/usr/local/lib/python3.6/dist-packages',
 '/usr/lib/python3/dist-packages',
 '/home/matt/.local/lib/python3.6/site-packages/IPython/extensions',
 '/home/matt/.ipython']